File: src/main/java/io/xol/chunkstories/api/content/Content.java
//
// This file is a part of the Chunk Stories API codebase
// Check out README.md for more information
// Website: http://chunkstories.xyz
//

package io.xol.chunkstories.api.content;

import java.util.Collection;
import java.util.Iterator;

import io.xol.chunkstories.api.voxel.materials.Material;
import org.slf4j.Logger;

import io.xol.chunkstories.api.GameContext;
import io.xol.chunkstories.api.animation.SkeletalAnimation;
import io.xol.chunkstories.api.content.mods.ModsManager;
import io.xol.chunkstories.api.entity.EntityDeclaration;
import io.xol.chunkstories.api.exceptions.net.UnknowPacketException;
import io.xol.chunkstories.api.item.ItemDeclaration;
import io.xol.chunkstories.api.mesh.MeshLibrary;
import io.xol.chunkstories.api.net.Packet;
import io.xol.chunkstories.api.net.PacketDefinition;
import io.xol.chunkstories.api.particles.ParticleTypeHandler;
import io.xol.chunkstories.api.rendering.voxel.VoxelRenderer;
import io.xol.chunkstories.api.voxel.Voxel;
import io.xol.chunkstories.api.voxel.textures.VoxelTexture;
import io.xol.chunkstories.api.world.World;
import io.xol.chunkstories.api.world.generator.WorldGenerator;

import javax.annotation.Nullable;

/** Encapsulates all the user-definable content available */
public interface Content {
	/** Returns the game context that owns this */
	GameContext getContext();

	/** Returns the ModManager that was used to load this content */
	ModsManager modsManager();

	/** Obtains an Asset using it's name string More advanced options for obtaining
	 * assets are avaible using the ModsManager class */
	@Nullable
	Asset getAsset(String assetName);

	/** Reloads everything. Warning: might not be appropriate as this will reload
	 * classes you will most likely have instanced into objects, causing weird
	 * errors if you do this while in a world. */
	void reload();

	Voxels voxels();

	interface Voxels {

		/** Returns the 'air' voxel ( No voxel data ) */
		Voxel air();

		@Nullable
		Voxel getVoxel(String voxelName);

		Iterator<Voxel> all();

		Content parent();

		VoxelTextures textures();

		interface VoxelTextures {

			VoxelTexture getVoxelTexture(String voxelTextureName);

			VoxelTexture getDefaultVoxelTexture();

			Iterator<VoxelTexture> all();

			Voxels parent();

			Logger logger();
		}

		VoxelMaterials materials();

		interface VoxelMaterials {

			Material getVoxelMaterial(String materialName);

			Iterator<Material> all();

			Material getDefaultMaterial();

			Content parent();

			Logger logger();
		}

		VoxelRenderer getDefaultVoxelRenderer();

		Logger logger();
	}

	ItemsDefinitions items();

	interface ItemsDefinitions {

		@Nullable
		ItemDeclaration getItemDefinition(String itemName);

		Iterator<ItemDeclaration> all();

		Content parent();

		Logger logger();
	}

	EntityDeclarations entities();

	interface EntityDeclarations {

		@Nullable
		EntityDeclaration getEntityDefinition(String entityName);

		Iterator<EntityDeclaration> all();

		Content parent();

		Logger logger();
	}

	ParticlesTypes particles();

	interface ParticlesTypes {

		@Nullable
		ParticleTypeHandler getParticleType(String string);

		Iterator<ParticleTypeHandler> all();

		Content parent();

		Logger logger();
	}

	PacketDefinitions packets();

	interface PacketDefinitions {

		@Nullable
		PacketDefinition getPacketByName(String name);

		PacketDefinition getPacketFromInstance(Packet packet) throws UnknowPacketException;

		Iterator<PacketDefinition> all();

		Content parent();

		Logger logger();
	}

	WorldGenerators generators();

	interface WorldGenerators {

		WorldGeneratorDefinition getWorldGenerator(String name);

		String getWorldGeneratorName(WorldGenerator generator);

		Iterator<WorldGeneratorDefinition> all();

		/** Contains the parameters stated in a 'generator' section of a .generators
		 * config file */
		interface WorldGeneratorDefinition extends Definition {
			String getName();

			/** Calls the constructor of whatever WorldGenerator you asked for */
			WorldGenerator createForWorld(World world);
		}

		Content parent();

		Logger logger();
	}

	LocalizationManager localization();

	interface LocalizationManager extends Translation {

		Collection<String> listTranslations();

		Content parent();

		void reload();

		void loadTranslation(String translationCode);

		Logger logger();
	}

	interface Translation {

		String getLocalizedString(String stringName);

		String localize(String text);
	}

	AnimationsLibrary getAnimationsLibrary();

	interface AnimationsLibrary {
		SkeletalAnimation getAnimation(String name);

		Content parent();

		void reloadAll();

		Logger logger();
	}

	MeshLibrary meshes();

	/** General logger about content */
	Logger logger();
}

File: src/main/java/io/xol/chunkstories/api/content/Content.kt
//
// This file is a part of the Chunk Stories API codebase
// Check out README.md for more information
// Website: http://chunkstories.xyz
//

package io.xol.chunkstories.api.content

import io.xol.chunkstories.api.GameContext
import io.xol.chunkstories.api.animation.Animation
import io.xol.chunkstories.api.content.mods.ModsManager
import io.xol.chunkstories.api.entity.EntityDefinition
import io.xol.chunkstories.api.exceptions.net.UnknowPacketException
import io.xol.chunkstories.api.graphics.Mesh
import io.xol.chunkstories.api.item.ItemDefinition
import io.xol.chunkstories.api.net.Packet
import io.xol.chunkstories.api.net.PacketDefinition
import io.xol.chunkstories.api.particles.ParticleTypeHandler
import io.xol.chunkstories.api.voxel.Voxel
import io.xol.chunkstories.api.voxel.materials.VoxelMaterial
import io.xol.chunkstories.api.voxel.textures.VoxelTexture
import io.xol.chunkstories.api.world.generator.WorldGenerator
import io.xol.chunkstories.api.world.generator.WorldGeneratorDefinition
import org.slf4j.Logger

/** Encapsulates all the user-definable content available  */
interface Content {
    /** Returns the game context that owns this  */
    val context: GameContext

    val animationsLibrary: AnimationsLibrary

    /** Returns the ModManager that was used to load this content  */
    fun modsManager(): ModsManager

    /** Obtains an Asset using it's name string More advanced options for obtaining
     * assets are avaible using the ModsManager class  */
    fun getAsset(assetName: String): Asset?

    /** Reloads everything. Warning: might not be appropriate as this will reload
     * classes you will most likely have instanced into objects, causing weird
     * errors if you do this while in a world.  */
    fun reload()

    fun voxels(): Voxels

    interface Voxels {

        /** Returns the 'air' voxel ( No voxel data )  */
        fun air(): Voxel

        fun getVoxel(voxelName: String): Voxel?

        fun all(): Iterator<Voxel>

        fun parent(): Content

        fun textures(): VoxelTextures

        interface VoxelTextures {

            val defaultVoxelTexture: VoxelTexture

            /** Looks for a voxel texture, if it fails it returns the default texture  */
            fun getVoxelTexture(voxelTextureName: String): VoxelTexture

            fun all(): Iterator<VoxelTexture>

            fun parent(): Voxels

            fun logger(): Logger
        }

        fun materials(): VoxelMaterials

        interface VoxelMaterials {

            val defaultMaterial: VoxelMaterial

            fun getVoxelMaterial(materialName: String): VoxelMaterial

            fun all(): Iterator<VoxelMaterial>

            fun parent(): Content

            fun logger(): Logger
        }

        fun logger(): Logger
    }

    fun items(): ItemsDefinitions

    interface ItemsDefinitions {

        fun getItemDefinition(itemName: String): ItemDefinition?

        fun all(): Iterator<ItemDefinition>

        fun parent(): Content

        fun logger(): Logger
    }

    fun entities(): EntityDefinitions

    interface EntityDefinitions {
        fun getEntityDefinition(entityName: String): EntityDefinition?

        fun all(): Iterator<EntityDefinition>

        fun parent(): Content

        fun logger(): Logger
    }

    fun particles(): ParticlesTypes

    interface ParticlesTypes {
        fun getParticleType(string: String): ParticleTypeHandler?

        fun all(): Iterator<ParticleTypeHandler>

        fun parent(): Content

        fun logger(): Logger
    }

    fun packets(): PacketDefinitions

    interface PacketDefinitions {
        fun getPacketByName(name: String): PacketDefinition?

        @Throws(UnknowPacketException::class)
        fun getPacketFromInstance(packet: Packet): PacketDefinition

        fun all(): Iterator<PacketDefinition>

        fun parent(): Content

        fun logger(): Logger
    }

    fun generators(): WorldGenerators

    interface WorldGenerators {
        fun getWorldGenerator(name: String): WorldGeneratorDefinition

        //fun getWorldGeneratorName(generator: WorldGenerator): String

        fun all(): Iterator<WorldGeneratorDefinition>

        fun parent(): Content

        fun logger(): Logger
    }

    fun localization(): LocalizationManager

    interface LocalizationManager : Translation {
        fun listTranslations(): Collection<String>

        fun parent(): Content

        fun reload()

        fun loadTranslation(translationCode: String)

        fun logger(): Logger
    }

    interface Translation {
        fun getLocalizedString(stringName: String): String

        fun localize(text: String): String
    }

    interface AnimationsLibrary {
        fun getAnimation(name: String): Animation

        fun parent(): Content

        fun reloadAll()

        fun logger(): Logger
    }

    interface Meshes {
        fun getMesh(meshName: String) : Mesh

        val defaultMesh: Mesh
    }

    fun meshes(): Meshes

    /** General logger about content  */
    fun logger(): Logger
}
